Client/Helpers/TabEventHandler.swift

```diff
@@ -0,0 +1,147 @@
+// empty methods, and `@objc optional` doesn't really work very well.
+extension TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {}
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {}
+}
+
+enum TabEventLabel: String {
+    case didLoadPageMetadata = "didLoadPageMetadata"
+    case didLoadFavicon = "didLoadFavicon"
+}
+
+enum TabEvent {
+    case didLoadPageMetadata(PageMetadata)
+    case didLoadFavicon(Favicon)
+    
+    var label: TabEventLabel {
+        switch self {
+        case .didLoadPageMetadata:
+            return .didLoadPageMetadata
+        case .didLoadFavicon:
+            return .didLoadFavicon
+        }
+    }
+    
+    func handle(_ tab: Tab, with handler: TabEventHandler) {
+        switch self {
+        case .didLoadPageMetadata(let metadata):
+            handler.tab(tab, didLoadPageMetadata: metadata)
+        case .didLoadFavicon(let favicon):
+            handler.tab(tab, didLoadFavicon: favicon)
+        }
+    }
+}
+
+// Hide some of the machinery away from the boiler plate above.
+////////////////////////////////////////////////////////////////////////////////////////
+extension TabEventLabel {
+    var name: Notification.Name {
+        return Notification.Name(self.rawValue)
+    }
+}
+
+extension TabEvent {
+    func notification(for tab: Any) -> Notification {
+        return Notification(name: label.name, object: tab, userInfo: ["payload": self])
+    }
+
+    /// Use this method to post notifications to any concerned listeners.
+    static func post(_ event: TabEvent, for tab: Any) {
+        NotificationCenter.default.post(event.notification(for: tab))
+    }
+}
+
+
+// These methods are used by TabEventHandler implementers.
+// Their usage remains consistent, even as we add more event types and handler methods.
+////////////////////////////////////////////////////////////////////////////////////////
+typealias TabObservers = [NSObjectProtocol]
+extension TabEventHandler {
+    /// Implementations of handles should use this method to register for events.
+    /// `TabObservers` should be preserved for unregistering later.
+    func registerFor(_ tabEvents: TabEventLabel..., queue: OperationQueue? = nil) -> TabObservers {
+        let center = NotificationCenter.default
+        return tabEvents.map { eventType in
+            center.addObserver(forName: eventType.name, object: nil, queue: queue) { notification in
+                guard let tab = notification.object as? Tab,
+                    let event = notification.userInfo?["payload"] as? TabEvent else {
+                        return
+                }
+                event.handle(tab, with: self)
+            }
+        }
+    }
+    
+    func unregister(_ observers: TabObservers) {
+        let center = NotificationCenter.default
+        observers.forEach { observer in
+            center.removeObserver(observer)
+        }
+    }
+}
```

Comment:
Could it be more generic:
```
  func registerFor(events: EventLabel...) -> NotificationObservers {
        let center = NotificationCenter.default
        return events.map { event in
            center.addObserver(forName: event.name, object: nil) { notification in
                guard let event = notification.userInfo?["payload"] as? NotificationEvent else {
                        return
                }
                event.handle(notification.object, with: self)
            }
        }
    }
    ```

Reply:
On further thought, no advantage to that. 
The PR code allows typecasting to happen here rather than `event.handle()`.